common/src/main/java/com/wynntils/wc/utils/lootrun/LootrunUtils.java

```diff
@@ -0,0 +1,542 @@
+import com.wynntils.mc.utils.McUtils;
+import com.wynntils.wc.utils.lootrun.objects.*;
+import it.unimi.dsi.fastutil.Pair;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ObjectObjectImmutablePair;
+import java.io.*;
+import java.text.DateFormat;
+import java.util.*;
+import javax.annotation.Nullable;
+import net.minecraft.ChatFormatting;
+import net.minecraft.client.Camera;
+import net.minecraft.client.gui.Font;
+import net.minecraft.client.gui.screens.inventory.ContainerScreen;
+import net.minecraft.client.player.LocalPlayer;
+import net.minecraft.client.renderer.LevelRenderer;
+import net.minecraft.client.renderer.MultiBufferSource;
+import net.minecraft.client.renderer.RenderType;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.core.BlockPos;
+import net.minecraft.network.chat.*;
+import net.minecraft.network.chat.Component;
+import net.minecraft.util.CubicSpline;
+import net.minecraft.util.FastColor;
+import net.minecraft.util.FormattedCharSequence;
+import net.minecraft.util.Mth;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class LootrunUtils {
+
+    public static final File LOOTRUNS = new File(WynntilsMod.MOD_STORAGE_ROOT, "lootruns");
+
+    private static final List<Integer> COLORS = List.of(
+            ChatFormatting.RED.getColor(),
+            ChatFormatting.GOLD.getColor(),
+            ChatFormatting.YELLOW.getColor(),
+            ChatFormatting.GREEN.getColor(),
+            ChatFormatting.BLUE.getColor(),
+            0x3f00ff,
+            ChatFormatting.DARK_PURPLE.getColor());
+
+    private static final int ACTIVE_COLOR = ChatFormatting.AQUA.getColor();
+    private static final int RECORDING_COLOR = 0xff0000;
+
+    public static LootrunInstance lootrun = null;
+    public static LootrunUncompiled uncompiled = null;
+    public static LootrunInstance recordingCompiled = null;
+    public static LootrunUncompiled recording = null;
+
+    public static RecordingInformation recordingInformation = null;
+
+    public static void register() {
+        WynntilsMod.getEventBus().register(LootrunUtils.class);
+    }
+
+    public static void unregister() {
+        WynntilsMod.getEventBus().register(LootrunUtils.class);
+    }
+
+    @SubscribeEvent
+    public static void render(RenderLevelLastEvent event) {
+        renderLootrun(event, lootrun, ACTIVE_COLOR);
+        renderLootrun(event, recordingCompiled, RECORDING_COLOR);
+    }
+
+    private static void renderLootrun(RenderLevelLastEvent event, LootrunInstance lootrun, int color) {
```

Comment:
This entire rendering function is not a thing of beauty. :( It is hard to read, no comments, way to long, and uses several booleans affecting flow, which makes it even harder to track.

At the very least, extract the points, chest and notes rendering parts to separate functions. After that, the points rendering will probably need a pass or two of cleanup to make it comprehensible. It seem to be doing some kind of "valid block" checking that can possibly be extracted in a separate function as well.

A simple rule of thumb is that a function should be at most a screen long. This is not always possible to keep, but this method is 150+ lines, which is way to much.